commit d2a7436c1c6144bbba2eb2a7b25db9b90515f0e6
Author: Stefan Liebler <stli@linux.ibm.com>
Date:   Tue Dec 18 13:57:24 2018 +0100

    S390: Refactor wmemset ifunc handling.
    
    The ifunc handling for wmemset is adjusted in order to omit ifunc
    if the minimum architecture level already supports newer CPUs by default.
    Unfortunately the c ifunc variant can't be omitted at all as it is used
    by the z13 ifunc variant as fallback if the pointers are not 4-byte aligned.
    Glibc internal calls will use the "newer" ifunc variant.
    
    ChangeLog:
    
            * sysdeps/s390/multiarch/Makefile
            (sysdep_routines): Remove wmemset variants.
            * sysdeps/s390/Makefile (sysdep_routines): Add wmemset variants.
            * sysdeps/s390/multiarch/ifunc-impl-list.c
            (__libc_ifunc_impl_list): Refactor ifunc handling for wmemset.
            * sysdeps/s390/multiarch/wmemset-c.c: Move to ...
            * sysdeps/s390/wmemset-c.c: ... here and adjust ifunc handling.
            * sysdeps/s390/multiarch/wmemset-vx.S: Move to ...
            * sysdeps/s390/wmemset-vx.S: ... here and adjust ifunc handling.
            * sysdeps/s390/multiarch/wmemset.c: Move to ...
            * sysdeps/s390/wmemset.c: ... here and adjust ifunc handling.
            * sysdeps/s390/ifunc-wmemset.h: New file.

diff --git a/sysdeps/s390/Makefile b/sysdeps/s390/Makefile
index fdfd1c605c28ddc7..f9a71276331b396a 100644
--- a/sysdeps/s390/Makefile
+++ b/sysdeps/s390/Makefile
@@ -76,5 +76,6 @@ sysdep_routines += wcslen wcslen-vx wcslen-c \
 		   wcsspn wcsspn-vx wcsspn-c \
 		   wcspbrk wcspbrk-vx wcspbrk-c \
 		   wcscspn wcscspn-vx wcscspn-c \
-		   wmemchr wmemchr-vx wmemchr-c
+		   wmemchr wmemchr-vx wmemchr-c \
+		   wmemset wmemset-vx wmemset-c
 endif
diff --git a/sysdeps/s390/ifunc-wmemset.h b/sysdeps/s390/ifunc-wmemset.h
new file mode 100644
index 0000000000000000..c9d1d17c3bfc7e9e
--- /dev/null
+++ b/sysdeps/s390/ifunc-wmemset.h
@@ -0,0 +1,53 @@
+/* wmemset variant information on S/390 version.
+   Copyright (C) 2018 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#if defined USE_MULTIARCH && IS_IN (libc)		\
+  && ! defined HAVE_S390_MIN_Z13_ZARCH_ASM_SUPPORT
+# define HAVE_WMEMSET_IFUNC	1
+#else
+# define HAVE_WMEMSET_IFUNC	0
+#endif
+
+#ifdef HAVE_S390_VX_ASM_SUPPORT
+# define HAVE_WMEMSET_IFUNC_AND_VX_SUPPORT HAVE_WMEMSET_IFUNC
+#else
+# define HAVE_WMEMSET_IFUNC_AND_VX_SUPPORT 0
+#endif
+
+#if defined HAVE_S390_MIN_Z13_ZARCH_ASM_SUPPORT
+# define WMEMSET_DEFAULT	WMEMSET_Z13
+/* The z13 ifunc variant is using the common code variant as fallback!  */
+# define HAVE_WMEMSET_C		1
+# define HAVE_WMEMSET_Z13	1
+#else
+# define WMEMSET_DEFAULT	WMEMSET_C
+# define HAVE_WMEMSET_C		1
+# define HAVE_WMEMSET_Z13	HAVE_WMEMSET_IFUNC_AND_VX_SUPPORT
+#endif
+
+#if HAVE_WMEMSET_C
+# define WMEMSET_C		__wmemset_c
+#else
+# define WMEMSET_C		NULL
+#endif
+
+#if HAVE_WMEMSET_Z13
+# define WMEMSET_Z13		__wmemset_vx
+#else
+# define WMEMSET_Z13		NULL
+#endif
diff --git a/sysdeps/s390/multiarch/Makefile b/sysdeps/s390/multiarch/Makefile
index 92e28dc45ddbae37..cc6dd7adb10ee8ad 100644
--- a/sysdeps/s390/multiarch/Makefile
+++ b/sysdeps/s390/multiarch/Makefile
@@ -1,6 +1,5 @@
 ifeq ($(subdir),wcsmbs)
-sysdep_routines += wmemset wmemset-vx wmemset-c \
-		   wmemcmp wmemcmp-vx wmemcmp-c
+sysdep_routines += wmemcmp wmemcmp-vx wmemcmp-c
 endif
 
 ifeq ($(subdir),iconvdata)
diff --git a/sysdeps/s390/multiarch/ifunc-impl-list.c b/sysdeps/s390/multiarch/ifunc-impl-list.c
index b5f55deb7faae9c4..7040959269c1612b 100644
--- a/sysdeps/s390/multiarch/ifunc-impl-list.c
+++ b/sysdeps/s390/multiarch/ifunc-impl-list.c
@@ -63,6 +63,7 @@
 #include <ifunc-wcspbrk.h>
 #include <ifunc-wcscspn.h>
 #include <ifunc-wmemchr.h>
+#include <ifunc-wmemset.h>
 
 /* Maximum number of IFUNC implementations.  */
 #define MAX_IFUNC	3
@@ -645,6 +646,18 @@ __libc_ifunc_impl_list (const char *name, struct libc_ifunc_impl *array,
 		)
 #endif /* HAVE_WMEMCHR_IFUNC  */
 
+#if HAVE_WMEMSET_IFUNC
+    IFUNC_IMPL (i, name, wmemset,
+# if HAVE_WMEMSET_Z13
+		IFUNC_IMPL_ADD (array, i, wmemset,
+				dl_hwcap & HWCAP_S390_VX, WMEMSET_Z13)
+# endif
+# if HAVE_WMEMSET_C
+		IFUNC_IMPL_ADD (array, i, wmemset, 1, WMEMSET_C)
+# endif
+		)
+#endif /* HAVE_WMEMSET_IFUNC  */
+
 #ifdef HAVE_S390_VX_ASM_SUPPORT
 
 # define IFUNC_VX_IMPL(FUNC)						\
@@ -653,8 +666,6 @@ __libc_ifunc_impl_list (const char *name, struct libc_ifunc_impl *array,
 			      __##FUNC##_vx)				\
 	      IFUNC_IMPL_ADD (array, i, FUNC, 1, __##FUNC##_c))
 
-  IFUNC_VX_IMPL (wmemset);
-
   IFUNC_VX_IMPL (wmemcmp);
 
 #endif /* HAVE_S390_VX_ASM_SUPPORT */
diff --git a/sysdeps/s390/multiarch/wmemset-c.c b/sysdeps/s390/wmemset-c.c
similarity index 59%
rename from sysdeps/s390/multiarch/wmemset-c.c
rename to sysdeps/s390/wmemset-c.c
index 1969cf93dcf08892..01e625496d8c2e4e 100644
--- a/sysdeps/s390/multiarch/wmemset-c.c
+++ b/sysdeps/s390/wmemset-c.c
@@ -16,22 +16,29 @@
    License along with the GNU C Library; if not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)
-# define WMEMSET  __wmemset_c
-
-# include <wchar.h>
-extern __typeof (__wmemset) __wmemset_c;
-# undef weak_alias
-# define weak_alias(name, alias)
-# ifdef SHARED
-#  undef libc_hidden_def
-#  define libc_hidden_def(name)					\
-  __hidden_ver1 (__wmemset_c, __GI___wmemset, __wmemset_c);
-#  undef libc_hidden_weak
-#  define libc_hidden_weak(name)					\
+#include <ifunc-wmemset.h>
+
+#if HAVE_WMEMSET_C
+# if HAVE_WMEMSET_IFUNC || HAVE_WMEMSET_Z13
+#  define WMEMSET WMEMSET_C
+
+#  undef weak_alias
+#  define weak_alias(name, alias)
+
+#  if defined SHARED && IS_IN (libc)
+#   undef libc_hidden_weak
+#   define libc_hidden_weak(name)
+#   undef libc_hidden_def
+#   if ! defined HAVE_S390_MIN_Z13_ZARCH_ASM_SUPPORT
+#    define libc_hidden_def(name)					\
+  __hidden_ver1 (__wmemset_c, __GI_wmemset, __wmemset_c)  __attribute__((weak)); \
   strong_alias (__wmemset_c, __wmemset_c_1);				\
-  __hidden_ver1 (__wmemset_c_1, __GI_wmemset, __wmemset_c_1);
-# endif /* SHARED */
+  __hidden_ver1 (__wmemset_c_1, __GI___wmemset, __wmemset_c_1);
+#   else
+#    define libc_hidden_def(name)
+#   endif
+#  endif
+# endif
 
 # include <wcsmbs/wmemset.c>
-#endif /* HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc) */
+#endif
diff --git a/sysdeps/s390/multiarch/wmemset-vx.S b/sysdeps/s390/wmemset-vx.S
similarity index 91%
rename from sysdeps/s390/multiarch/wmemset-vx.S
rename to sysdeps/s390/wmemset-vx.S
index 0c2f6337b0e554ec..4b6050b5accd732b 100644
--- a/sysdeps/s390/multiarch/wmemset-vx.S
+++ b/sysdeps/s390/wmemset-vx.S
@@ -16,7 +16,8 @@
    License along with the GNU C Library; if not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)
+#include <ifunc-wmemset.h>
+#if HAVE_WMEMSET_Z13
 
 # include "sysdep.h"
 # include "asm-syntax.h"
@@ -38,7 +39,7 @@
    -v17,v18,v19=copy of v16 for vstm
    -v31=saved dest for return
 */
-ENTRY(__wmemset_vx)
+ENTRY(WMEMSET_Z13)
 	.machine "z13"
 	.machinemode "zarch_nohighgprs"
 
@@ -137,6 +138,17 @@ ENTRY(__wmemset_vx)
 	br	%r14
 .Lfallback:
 	srlg	%r4,%r4,2	/* Convert byte-count to character-count.  */
-	jg	__wmemset_c
-END(__wmemset_vx)
-#endif /* HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc) */
+	jg	WMEMSET_C
+END(WMEMSET_Z13)
+
+# if ! HAVE_WMEMSET_IFUNC
+strong_alias (WMEMSET_Z13, __wmemset)
+weak_alias (__wmemset, wmemset)
+# endif
+
+# if defined HAVE_S390_MIN_Z13_ZARCH_ASM_SUPPORT \
+	&& defined SHARED && IS_IN (libc)
+strong_alias (WMEMSET_Z13, __GI___wmemset)
+weak_alias (WMEMSET_Z13, __GI_wmemset)
+# endif
+#endif
diff --git a/sysdeps/s390/multiarch/wmemset.c b/sysdeps/s390/wmemset.c
similarity index 70%
rename from sysdeps/s390/multiarch/wmemset.c
rename to sysdeps/s390/wmemset.c
index 149b4814708d820a..6118754d1df71948 100644
--- a/sysdeps/s390/multiarch/wmemset.c
+++ b/sysdeps/s390/wmemset.c
@@ -16,7 +16,9 @@
    License along with the GNU C Library; if not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)
+#include <ifunc-wmemset.h>
+
+#if HAVE_WMEMSET_IFUNC
 # define wmemset __redirect_wmemset
 # define __wmemset __redirect___wmemset
 # include <wchar.h>
@@ -24,9 +26,18 @@
 # undef __wmemset
 # include <ifunc-resolve.h>
 
-s390_vx_libc_ifunc_redirected (__redirect___wmemset, __wmemset)
-weak_alias (__wmemset, wmemset)
+# if HAVE_WMEMSET_C
+extern __typeof (__redirect___wmemset) WMEMSET_C attribute_hidden;
+# endif
 
-#else
-# include <wcsmbs/wmemset.c>
-#endif /* !(defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)) */
+# if HAVE_WMEMSET_Z13
+extern __typeof (__redirect___wmemset) WMEMSET_Z13 attribute_hidden;
+# endif
+
+s390_libc_ifunc_expr (__redirect___wmemset, __wmemset,
+		      (HAVE_WMEMSET_Z13 && (hwcap & HWCAP_S390_VX))
+		      ? WMEMSET_Z13
+		      : WMEMSET_DEFAULT
+		      )
+weak_alias (__wmemset, wmemset)
+#endif
